File: BackEnd/Utils.Notifications/NotificationsRealtime.cs
using Ibt.Ortc.Api.Extensibility;
using System;

namespace Utils.Notifications
{
    public class NotificationsRealtime : INotifications
    {

        private OrtcClient _ortcClient;
        private string _channelName;

        public NotificationsRealtime()
        {
            var api = new Ibt.Ortc.Api.Ortc();
            IOrtcFactory factory = api.LoadOrtcFactory("IbtRealTimeSJ");
            _ortcClient = factory.CreateClient();
            _ortcClient.ClusterUrl = "http://ortc-developers.realtime.co/server/2.1/";
            _ortcClient.ConnectionMetadata = "testConnectionMetadata";
            // Se asignan métodos de callback para manejar conexión exitosa
            // y suscripción exitosa a canales.
            _ortcClient.OnConnected += new OnConnectedDelegate(OrtcOnConnected);
            _ortcClient.OnSubscribed += new OnSubscribedDelegate(OrtcOnSubscribed);
            // Se establece clave de aplicación y token de autenticación.
            _ortcClient.Connect("HDzwHh", "myToken");
            // Se setea channelName y message temporales.
            _channelName = "";
        }

        //TODO eliminar esto en principio.
        public void CreateInstance()
        {

        }

        /// <summary>
        /// Método utilizado para enviar un mensaje a un canal en particular.
        /// </summary>
        /// <param name="channelName">Nombre de canal.</param>
        /// <param name="message">Mensaje.</param>
        public void SendMessage(string channelName, string message)
        {
            // Si el nombre de canal es distinto al actual, se desuscribe del
            // canal actual y se genera la suscripción al nuevo.
            if (!_channelName.Equals(channelName, StringComparison.Ordinal))
            {
                PrintDebugMessage(@"No se está suscrito a ese canal.");
            } else
            {
                _ortcClient.Send(channelName, message);
            }
        }



        /// <summary>
        /// Se genera la suscripción por parte del cliente a un canal de comunicación
        /// en concreto.
        /// </summary>
        /// <param name="chanelName">Nombre del canal.</param>
        public void SubscribeChanel(string channelName)
        {
            if (_channelName.Equals(channelName, StringComparison.Ordinal))
            {
                PrintDebugMessage("Ya existe una suscripción al canal " 
                    + channelName + ".");
            }
            else
            {
                // Si se está suscrito a un canal, se anula la suscripción
                if (!_channelName.Equals("", StringComparison.Ordinal))
                {
                    PrintDebugMessage(@"Debe desuscribirse de un canal antes de intentar
                        suscribirse a uno nuevo.");
                }
                else
                {
                    _channelName = channelName;
                    _ortcClient.Subscribe(_channelName, true, OnMessageCallback);
                }
            }
        }

        /// <summary>
        /// Se elimina la suscripción de un cliente a un canal en particular.
        /// </summary>
        /// <param name="channelName">Nombre del canal.</param>
        public void UnsubscribeChanel(string channelName)
        {
            _channelName = "";
            _ortcClient.Unsubscribe(channelName);
        }

        /// <summary>
        /// Método llamado cuando el cliente de Realtime Framework consigue
        /// establecer la conexión.
        /// </summary>
        /// <param name="sender"></param>
        private void OrtcOnConnected(object sender)
        {
            _ortcClient.Subscribe(_channelName, true, OnMessageCallback);
        }

        /// <summary>
        /// Método llamado cuando se recibe un mensaje por el canal al que
        /// el cliente está suscrito, sirve para manejarlo.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="channel">Canal de comunicación utilizado.</param>
        /// <param name="message">Mensaje recibido.</param>
        private void OnMessageCallback(object sender, string channel, string message)
        {
            Console.WriteLine(channel + " > " + message);
        }

        /// <summary>
        /// Método llamado cuando el cliente consiguió subscribirse
        /// exitosamente a un canal de comunicación.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="channel"></param>
        private void OrtcOnSubscribed(object sender, string channel)
        {
            // Sending Hello World message
            _ortcClient.Send(channel, "Backend de Realtime Framework conectado.");
        }

        private void PrintDebugMessage(string msg)
        {
            Console.WriteLine(msg);
        }

    }
}

File: FrontEnd/MemsysAndroid/app/src/main/java/pis16/memsysandroid/notifications/NotificationsRealtime.java
package pis16.memsysandroid.notifications;

import ibt.ortc.api.Ortc;
import ibt.ortc.extensibility.OnConnected;
import ibt.ortc.extensibility.OnMessage;
import ibt.ortc.extensibility.OrtcClient;
import ibt.ortc.extensibility.OrtcFactory;
import pis16.memsysandroid.notifications.interfaces.INotificationReceiver;
import pis16.memsysandroid.notifications.interfaces.INotifications;
import pis16.memsysandroid.ui.MainActivity;

/**
 * Created by marccio on 29-Aug-16.
 */
public class NotificationsRealtime implements INotifications {

    private OrtcClient client;
    INotificationReceiver caller;
    String appKey = "HDzwHh";
    String token = "testToken";
    String googleProjectId = "757215377367";

    public NotificationsRealtime(final INotificationReceiver caller) {
        this.caller = caller;
        OrtcFactory factory;
        Ortc ortc = new Ortc();
        try {
            factory = ortc.loadOrtcFactory("IbtRealtimeSJ");
            client = factory.createClient();
//            client.setApplicationContext(caller.getApplicationContext());
            client.setGoogleProjectId(googleProjectId);
            client.setClusterUrl("http://ortc-developers.realtime.co/server/2.1/");
            client.onConnected = new OnConnected() {
                @Override
                public void run(OrtcClient sender) {
                    System.out.println("Cliente en Android conectado.");
                }
            };
            client.connect(appKey, token);
        } catch (Exception e) {
            System.out.println(String.format("Realtime Error: %s", e.toString()));
        }
    }

    @Override
    public void SubscribeChannel(String channelName) {
        client.subscribe(channelName, true,
                new OnMessage() {
                    // Esta función maneja los mensajes recibidos en el canal
                    // al que el cliente esté suscrito.
                    public void run(OrtcClient sender, String channel, String message) {
                        // Received a message
                        System.out.println(channel + " > " + message);
                        caller.onMessageReceived(message);
                    }
                });
        System.out.println("Cliente suscrito al canal " + channelName);
    }

    @Override
    public void UnsubscribeChannel(String channelName) {
        client.unsubscribe(channelName);
    }


    @Override
    public  void SendMessage(String channelName, String message) {

    }

    @Override
    public boolean isConnected() {
        return client.getIsConnected();
    }


}

